Sources/ShopifyCheckoutKit/AnalyticsEvents.swift

```diff
@@ -0,0 +1,4594 @@
+        searchSubmitted: PixelEventsSearchSubmitted?? = nil
+    ) -> PixelEvents {
+        return PixelEvents(
+            cartViewed: cartViewed ?? self.cartViewed,
+            checkoutAddressInfoSubmitted: checkoutAddressInfoSubmitted ?? self.checkoutAddressInfoSubmitted,
+            checkoutCompleted: checkoutCompleted ?? self.checkoutCompleted,
+            checkoutContactInfoSubmitted: checkoutContactInfoSubmitted ?? self.checkoutContactInfoSubmitted,
+            checkoutShippingInfoSubmitted: checkoutShippingInfoSubmitted ?? self.checkoutShippingInfoSubmitted,
+            checkoutStarted: checkoutStarted ?? self.checkoutStarted,
+            collectionViewed: collectionViewed ?? self.collectionViewed,
+            pageViewed: pageViewed ?? self.pageViewed,
+            paymentInfoSubmitted: paymentInfoSubmitted ?? self.paymentInfoSubmitted,
+            productAddedToCart: productAddedToCart ?? self.productAddedToCart,
+            productRemovedFromCart: productRemovedFromCart ?? self.productRemovedFromCart,
+            productVariantViewed: productVariantViewed ?? self.productVariantViewed,
+            productViewed: productViewed ?? self.productViewed,
+            searchSubmitted: searchSubmitted ?? self.searchSubmitted
+        )
+    }
+
+    func jsonData() throws -> Data {
+        return try newJSONEncoder().encode(self)
+    }
+
+    func jsonString(encoding: String.Encoding = .utf8) throws -> String? {
+        return String(data: try self.jsonData(), encoding: encoding)
+    }
+}
+
+/// The `cart_viewed` event logs an instance where a customer visited the cart
+/// page
+// MARK: - PixelEventsCartViewed
+struct PixelEventsCartViewed: Codable {
+    /// The client-side ID of the customer, provided by Shopify
+    let clientID: String?
+    let context: Context?
+    let data: PixelEventsCartViewedData?
+    /// The ID of the customer event
+    let id: String?
+    /// The name of the customer event
+    let name: PixelEventsCartViewedName?
+    /// The timestamp of when the customer event occurred, in [ISO
+    /// 8601](https://en.wikipedia.org/wiki/ISO_8601) format
+    let timestamp: String?
+
+    enum CodingKeys: String, CodingKey {
+        case clientID = "clientId"
+        case context, data, id, name, timestamp
+    }
+}
+
+// MARK: PixelEventsCartViewed convenience initializers and mutators
+
+extension PixelEventsCartViewed {
+    init(data: Data) throws {
+        self = try newJSONDecoder().decode(PixelEventsCartViewed.self, from: data)
+    }
+
+    init(_ json: String, using encoding: String.Encoding = .utf8) throws {
+        guard let data = json.data(using: encoding) else {
+            throw NSError(domain: "JSONDecoding", code: 0, userInfo: nil)
+        }
+        try self.init(data: data)
+    }
+
+    init(fromURL url: URL) throws {
+        try self.init(data: try Data(contentsOf: url))
+    }
+
+    func with(
+        clientID: String?? = nil,
+        context: Context?? = nil,
```

Comment:
I think events each have a context, id, name, timestamp with the same type.

Can we introduce an interface or supertype?